### DISCUSSION TITLE
Is it possible to turn off auto-closing-item completion?

### DISCUSSION THREAD

**Discussion by arifd** [zed-industries/zed#21872] [Feature Requests]
Zed is trying to be clever, and I get it, I start an open brace or pipe, and it tries to place the closing pipe where it "makes sense".

However, it almost never knows my intention, so it puts it in the wrong place, and now, for every open brace, I type, I have to spend time trying to figure out where zed broke my code.

It's really frustrating and I'd love to just not have it do it at all.

Thank you!!

**Comment by SomeoneToIgnore** [comment]
It's not Zed but the language server in action judging from the sympthoms, what is the language in question?

**Reply by arifd** [reply]
Ouch, okay. I'm using Rust, I guess with rust-analyzer, then? Would you know how to disable it? 8/10 it does not do what I want and the other 2 times I have to manually check anyway! 😓

**Reply by SomeoneToIgnore** [reply]
Most probably it's OnTypeFormating server requests from r-a.
For `|`, the fix is expected in the next Monday update of r-a: https://github.com/rust-lang/rust-analyzer/pull/18647

Meanwhile, `use_on_type_format: false` can be put in settings to disable this:
https://github.com/zed-industries/zed/blob/010fdeefb75ad7815fc18d2ed5bf7f2d52ee1ce2/assets/settings/default.json#L171-L173 then.

Or, configured on LSP side which symbols are ignored: 
https://github.com/rust-lang/rust-analyzer/issues/18641
using examples from https://zed.dev/docs/languages/rust

**Reply by arifd** [reply]
Oh great, yeah hahaha the pipes were the last straw actually! 😀 

But I'm not sure how to translate that feature to a zed setting.

in vscode I already have this:

```json
{
  "rust-analyzer.server.extraEnv": {
    "CARGO_TARGET_DIR": "target/analyzer",
    "SOME_OTHER_ENV": "2099-01-01T00:00:00Z"
  },
  "rust-analyzer.check.extraArgs": ["--target-dir=target/analyzer"],
  "rust-analyzer.cargo.buildScripts.rebuildOnSave": true,
  "rust-analyzer.diagnostics.enable": false
}
```

And it works great (especially the custom target dir, to stop RA and clippy blocking each other).

In zed documentation, I am told to write it like this:

```json
"lsp": {
    "rust-analyzer": {
      "initialization_options": {
        "rust": {
          "analyzerTargetDir": "target/analyzer"
        }
      }
    },
}
```

Which makes me think zed is basically parsing it and forwarding it onto RA. But I don't want to learn a zed settings protocol and a RA settings protocol.

In any case, RA and clippy are still blocking eachother in zed, whereas they don't in vscode.

Is there a way to just mirror the vscode options directly in zed. I.e. would this just work?

```json
"lsp": {
  "rust-analyzer.server.extraEnv": {
    "CARGO_TARGET_DIR": "target/analyzer",
    "SOME_OTHER_ENV": "2099-01-01T00:00:00Z"
  },
  "rust-analyzer.check.extraArgs": ["--target-dir=target/analyzer"],
  "rust-analyzer.cargo.buildScripts.rebuildOnSave": true,
  "rust-analyzer.diagnostics.enable": false
}
```

And then finally, where and how to add `"use_on_type_format": false` to it?

Thanks so much, and sorry if this is RTFM material, but i have tried to consume the documenation but i think there's more of a conceptual gap, for me, than an informational one.

**Reply by SomeoneToIgnore** [reply]
There's `zed: open settings` command in Zed (cmd+, on mac) which will open user settings.
It should have the same structure as the default settings (https://github.com/zed-industries/zed/blob/010fdeefb75ad7815fc18d2ed5bf7f2d52ee1ce2/assets/settings/default.json#L171-L173, `zed: open default settings` command in Zed) but contain only things that are different from the defaults.

There are two logical places to configure things:

* if you like the language-related behavior in general but prefer nuances, that usually goes to under `lsp` section (e.g. turn off the pipes). The zed.dev link shows certain fields conversion, which look like if the vscode items were split by `.` and used nested json objects instead.

* turn off in the editor entirely, then a top-level section in the user settings with `"use_on_type_format": false`

**Reply by arifd** [reply]
but if this option is simply forwarded to RA, is there a way that I can tell RA not to do it, so that my settings are indepdendent of the editor (in this case, zed)?

And can i just plonk straight vscode RA settings into zed, by nesting them in "lsp"?

**Reply by arifd** [reply]
Great thank you.
